### ISSUE TITLE
[FEATURE REQUEST] SQLServerBulkCopy OffsetDateTime Support

### ISSUE THREAD

**Issue by pbcornelius** [microsoft/mssql-jdbc#1363]
As the title suggests: implementation of OffsetDateTime support in SQLServerBulkCopy. It should be fairly simple to create a SQL Server DatetimeOffset from a Java OffsetDateTime in https://github.com/microsoft/mssql-jdbc/blob/dev/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java#L2828, and then pass that along.

Unfortunately this cannot easily be circumvented on the source side (for instance, by performing the conversion in the source ResultSet), because SQLServerBulkCopy casts the source to a SQLServerResultSet for type DatetimeOffset.

**Comment by peterbae** [comment]
Hi @pbcornelius, thanks for contacting us. Users are expected to provide a `DateTimeOffset` object along with providing `microsoft.sql.DateTimeOffset` as the column type when performing bulk copy (you can refer to the type mapping table [here](https://docs.microsoft.com/en-us/sql/connect/jdbc/using-basic-data-types?view=sql-server-2017)). OffsetDateTime and DateTimeOffset actually have subtle differences that set them apart from one another, such as the application of timezones. Could you explain your bulk copy scenario in further detail? Thanks.

**Comment by pbcornelius** [comment]
I'm coming from H2 JDBC and going to MSSQL.

I want to provide a `DateTimeOffset` object, but I don't think I can do that coming from a non-SQL Server `ResultSet` because of this code in `SQLServerBulkCopy.readColumnFromResultSet(...)`:

```
                case microsoft.sql.Types.DATETIMEOFFSET:
                    // We can safely cast the result set to a SQLServerResultSet as the DatetimeOffset type is only
                    // available in the JDBC driver.
                    return ((SQLServerResultSet) sourceResultSet).getDateTimeOffset(srcColOrdinal);
```

The assumption in this code is that any `ResultSet` capable of providing `DateTimeOffset` is actually a `SQLServerResultSet`. Therefore, any result set coming from non-MSSQL cannot provide DateTimeOffset to SQLServerBulkCopy. Even if my custom `sourceResultSet` did implement `getDateTimeOffset(...)`, it wouldn't run because it would fail the cast to `SQLServerResultSet`. I'm guessing, but I can see two solutions: (1) either implement the conversion of OffsetDateTime to DateTimeOffset once for everyone in the SQLServerBulkCopy, or (2) use getObject(column, DateTimeOffset.class) in the above code for non-MSSQL result sets.

**Comment by peterbae** [comment]
I see. I agree that the driver should avoid casting to SQLServerResultSet and use getObject instead to facilitate using non-MSSQL result sets, so I can definitely make that change. I'll create a PR for this after doing some testing to see if this change won't break anything.

**Comment by peterbae** [comment]
Merged PR #1365. Closing issue.
